assets/docs/configuration/targets/http_response_handler.hcl

```diff
@@ -0,0 +1,60 @@
+target {
+  use "http" {
+
+    response_handler {
+
+     // everything is fine, we can ack our data from source.
+     // Can such configuration be useful or it's overkill and we should commit to 2** all the time?
+      success = [
+        //we can have concrete status or wildcard using *
+        { http_status: ["2**"]}
+      ] 
+
+      // data is for some reason not accepted by target, there is no point of retrying. Just send it to bad/failed target and unblock processing.
+      invalid/badrow = [
+
+        //first rule...
+        { http_status: ["41*"]},
+
+        //second rule...
+        { 
+          http_status: ["499", "424"],
+
+          //not only http status but check if response body matches. Extract some part of body (e.g. error code) and check if it's equal/not equal to value.
+          body {
+            path: "some jq here??"
+            equal/not_equal: "some value"
+          }
+        }
+      ]
+
+      // For these retryable errors (assuming we have health check):
+      // - set app status as unhealthy for each failed retry
```

Comment:
Do the streaming loaders have this concept of being set as unhealthy after a failed retry? Are there very few retries?

Would like to understand/think through this bit together. We have an app here that makes thousands of requests per second, and depending on the target, 10-15% of them might get rejected and retried.

If we make it configurable this might not be a problem but would like to understand it better regardless. :)

Reply:
> Do the streaming loaders have this concept of being set as unhealthy after a failed retry? Are there very few retries?

Yes, kind of. But in loaders we do that for synchronous actions like init table, create columns, open channel. If it's transient error or setup error, we mark app as unhealthy.

Here I think it's a bit different. We can have a lot of concurrent writes to HTTP target, right? So we may end up in situation like you described - 15% of request are retried (some transient issue) and they mark app as unhealthy. The rest 85% are OK and they set it back to healthy. 

Sooo we may have a situation when we flip back and forth health toggle and we are unlucky with Kubernetes periodic liveness probe when it sees only unhealthy state....and kills the pod, even though overall everything looks pretty good?

It would make sense to make it unhealthy only for those persistent setup-like issues. I could add configuration option like `toggle_health` to retry strategy and set to our default `setup` one as `true`. Or instead of boolean use threshold, like `unhealthy_after` with number of retries after which we mark app as unhealthy....?

Reply:
> Here I think it's a bit different. We can have a lot of concurrent writes to HTTP target, right? So we may end up in situation like you described - 15% of request are retried (some transient issue) and they mark app as unhealthy. The rest 85% are OK and they set it back to healthy.

Pretty much - what's more we can expect to sometimes be in that state for a long time (eg. when being rate limited on a spike of traffic), and we actually rely on getting failures in those scenarios. So something feels somewhat odd about considering individual failures to mean 'unhealthy' status.

> Sooo we may have a situation when we flip back and forth health toggle and we are unlucky with Kubernetes periodic liveness probe when it sees only unhealthy state....and kills the pod, even though overall everything looks pretty good?

I'm not sure how much of a real concern this scenario is, becuase we'd likely have many more successes than failures. But even still something feels odd about the design in this scenario.

> It would make sense to make it unhealthy only for those persistent setup-like issues. I could add configuration option like toggle_health to retry strategy and set to our default setup one as true. Or instead of boolean use threshold, like unhealthy_after with number of retries after which we mark app as unhealthy....?

If we're only concerned with setup issues, then one option is a global flag to mark whether _any_ request has been successful would work - on any success we'd mark it as true, on failure we wouldn't mark it as false but would mark the app as unhealthy. 

But I'm not sure about this - would you mind jotting down a quick synopsis of this scenario? (maybe add a details section to the existing doc - but anywhere is fine)

I feel like getting a grip on the desired behaviour for this type of app might lead us to a more confident solution. :)